src/database/import.ts

```diff
@@ -0,0 +1,166 @@
+import { Client } from "../apiv2";
+import { FirebaseError } from "../error";
+
+const JSONStream = require("JSONStream");
+
+const MAX_CHUNK_SIZE = 1024 * 1024;
+const CONCURRENCY_LIMIT = 5;
+
+type Data = {
+  json: { [key: string]: any } | string | number | boolean;
+  pathname: string;
+};
+
+type ChunkedData = {
+  chunks: Data[] | null;
+  size: number;
+};
+
+/**
+ * Imports JSON data to a given RTDB instance.
+ *
+ * The data is parsed and chunked into subtrees of ~1 MB, to be subsequently written in parallel.
+ */
+export default class DatabaseImporter {
+  private jsonPath: string;
+  private client: Client;
+  private limit = pLimit(CONCURRENCY_LIMIT);
+
+  constructor(
+    private dbUrl: URL,
+    private inStream: NodeJS.ReadableStream,
+    dataPath: string,
+    private chunkSize = MAX_CHUNK_SIZE
+  ) {
+    this.jsonPath = this.computeJsonPath(dataPath);
+    this.client = new Client({ urlPrefix: dbUrl.origin, auth: true });
+  }
+
+  /**
+   * Writes the chunked data to RTDB. Any existing data at the specified location will be overwritten.
+   */
+  async execute(): Promise<any> {
+    await this.checkLocationIsEmpty();
+    return this.readAndWriteChunks(this.inStream);
+  }
+
+  private async checkLocationIsEmpty(): Promise<void> {
+    const response = await this.client.request({
+      method: "GET",
+      path: this.dbUrl.pathname + ".json",
+      queryParams: { shallow: "true" },
+    });
+
+    if (response.body) {
+      throw new FirebaseError(
+        "Importing is only allowed for an empty location. Delete all data by running " +
+          clc.bold(`firebase database:remove ${this.dbUrl.pathname} --disable-triggers`) +
+          ", then rerun this command.",
+        { exit: 2 }
+      );
+    }
+  }
+
+  private readAndWriteChunks(inStream: NodeJS.ReadableStream): Promise<any> {
+    const { dbUrl } = this;
+    const chunkData = this.chunkData.bind(this);
+    const writeChunk = this.writeChunk.bind(this);
+    const getJoinedPath = this.joinPath.bind(this);
+
+    const readChunks = new stream.Transform({ objectMode: true });
+    readChunks._transform = function (chunk: { key: string; value: any }, _, done) {
+      const data = { json: chunk.value, pathname: getJoinedPath(dbUrl.pathname, chunk.key) };
+      const chunkedData = chunkData(data);
+      const chunks = chunkedData.chunks || [data];
+      chunks.forEach((chunk: Data) => this.push(chunk));
+      done();
+    };
+
+    const writeChunks = new stream.Transform({ objectMode: true });
+    writeChunks._transform = async function (chunk: Data, _, done) {
+      const res = await writeChunk(chunk);
+      this.push(res);
+      done();
+    };
+
+    return new Promise((resolve, reject) => {
+      const responses: any[] = [];
+      inStream
+        .pipe(JSONStream.parse(this.jsonPath))
+        .on("error", (err: any) =>
+          reject(
+            new FirebaseError("Invalid data; couldn't parse JSON object, array, or value.", {
```

Comment:
Could you test this error case to see how it appears to users? I want to make sure the original error is visible to them without the --debug flag

Reply:
Fixed:
```
Error: Invalid data; couldn't parse JSON object, array, or value. Unexpected COLON(":") in state COMMA
```